torch/utils/_as_strided.py

```diff
@@ -0,0 +1,492 @@
+                break
+
+        if not target_dims_for_this_canonical:
+            # No target dims map to this canonical dim, skip it
+            # (This dimension gets "lost" in the transformation)
+            continue
+
+        # Process target_dims left to right (largest stride first, already sorted)
+        for i, (original_pos, target_size, target_stride) in enumerate(
+            target_dims_for_this_canonical
+        ):
+            is_last = i == len(target_dims_for_this_canonical) - 1
+
+            if not is_last:
+                # unflatten(a*b, (a, b)) gives strides (b*canonical_stride, canonical_stride)
+                # To get first dim stride = target_stride: b * canonical_stride = target_stride
+                # So b = target_stride / canonical_stride
+                next_dim_size = target_stride // canonical_stride
+
+                needed_size = target_size * next_dim_size
+
+                # Check if we have enough size
+                if result.size(current_dim) < needed_size:
+                    return NotImplemented
+
+                # Narrow and unflatten
+                if result.size(current_dim) != needed_size:
+                    result = result.narrow(current_dim, 0, needed_size)
+
+                result = result.unflatten(current_dim, (target_size, next_dim_size))
+
+            else:
+                # Last target dim
+                # Its stride must match the canonical stride
+                if target_stride != canonical_stride:
+                    return NotImplemented
+
+                # Narrow to final size
+                if result.size(current_dim) < target_size:
+                    return NotImplemented
+                if result.size(current_dim) != target_size:
+                    result = result.narrow(current_dim, 0, target_size)
+
+            current_dim += 1
+
+    return result
+
+
+def as_strided_via_views(
+    tensor: torch.Tensor,
+    size: tuple[int, ...],
+    stride: tuple[int, ...],
+    storage_offset: int = 0,
+) -> torch.Tensor | types.NotImplementedType:
+    """
+    Attempt to reconstruct a sequence of view operations that would produce
+    the target size/stride/offset from the source tensor.
+
+    Args:
+        tensor: Source tensor
+        size: Target size
+        stride: Target stride
+        storage_offset: Target storage offset (default: 0)
+
+    Returns:
+        A view of the tensor with the target size/stride/offset, or NotImplemented
+        if it cannot be achieved via simple view operations.
+    """
+    import math
+
+    # NB: When strides are insignificant, we don't reproduce them exactly, as
+    # this would make the algorithm a lot more complicated for no reason
+
+    target_numel = math.prod(size)
+
+    # Easy case 1: source tensor has numel==0
+    if tensor.numel() == 0:
+        if target_numel != 0:
+            return NotImplemented
+        return tensor.view(size)
```

Comment:
Do we need strides to be correct for this one as well?

Reply:
I was kind of hoping to first ship a version of this patch that doesn't do insignificant strides correctly, and then have a separate work item to fix the stride situation. Not sure if that is too much for you though :P